keras/src/layers/rnn/gru_test.py

```diff
@@ -400,3 +400,33 @@ def test_legacy_implementation_argument(self):
             tpu_atol=1e-3,
             tpu_rtol=1e-3,
         )
+
+    def test_stateful_with_symbolic_batch_size(
+        self,
+    ):
+        from keras.src import backend
+
+        if backend.backend() != "tensorflow":
+            self.skipTest(
+                "Test only applicable to fixing a bug for TensorFlow backend."
+            )
+
+        layer = layers.GRU(
+            5,
+            stateful=True,
+        )
+
+        x_concrete = np.ones((2, 10, 10), dtype=np.float32)
+        _ = layer(x_concrete, training=True)
+        import tensorflow as tf
+
+        @tf.function(
+            input_signature=[
+                tf.TensorSpec(shape=(None, 10, 10), dtype=tf.float32)
+            ]
+        )
+        def f(x):
+            return layer(x, training=True)
+
+        y = f(x_concrete)
+        self.assertEqual(y.shape, (2, 5))
```

Comment:
I think this test can be re-written this way:

```python
model = models.Sequential([layers.GRU(5, stateful=True)])
model.jit_compile = True

x = np.ones((5, 10, 10), dtype=np.float32)
model(x)

y = model.predict(x, batch_size=2)
self.assertEqual(y.shape, (2, 5))
```

Which I like better because you may be able to run it with other backends since there's no TF specific code. The trick is that `predict` will use a dynamic batch size because it has to.

Let me know if this doesn't work or it doesn't exercise the issue you're fixing.

Reply:
I actually had a similar idea and the original test I had was written to that effect. Unfortunately it didn't exercise what's being tested (as in, I could revert the change and the test still passes)